File: crates/codegraph-core/src/extractors/java.rs
use super::helpers::*;
use super::SymbolExtractor;
use crate::cfg::build_function_cfg;
use crate::complexity::compute_all_metrics;
use crate::types::*;
use tree_sitter::{Node, Tree};

pub struct JavaExtractor;

impl SymbolExtractor for JavaExtractor {
    fn extract(&self, tree: &Tree, source: &[u8], file_path: &str) -> FileSymbols {
        let mut symbols = FileSymbols::new(file_path.to_string());
        walk_tree(&tree.root_node(), source, &mut symbols, match_java_node);
        walk_ast_nodes_with_config(&tree.root_node(), source, &mut symbols.ast_nodes, &JAVA_AST_CONFIG);
        walk_tree(&tree.root_node(), source, &mut symbols, match_java_type_map);
        symbols
    }
}

// ── Type inference helpers ──────────────────────────────────────────────────

fn extract_java_type_name<'a>(type_node: &Node<'a>, source: &'a [u8]) -> Option<&'a str> {
    if type_node.kind() == "generic_type" {
        type_node.child(0).map(|n| node_text(&n, source))
    } else {
        Some(node_text(type_node, source))
    }
}

fn match_java_type_map(node: &Node, source: &[u8], symbols: &mut FileSymbols, _depth: usize) {
    match node.kind() {
        "local_variable_declaration" => {
            if let Some(type_node) = node.child_by_field_name("type") {
                if let Some(type_name) = extract_java_type_name(&type_node, source) {
                    for i in 0..node.child_count() {
                        if let Some(child) = node.child(i) {
                            if child.kind() == "variable_declarator" {
                                if let Some(name_node) = child.child_by_field_name("name") {
                                    push_type_map_entry(
                                        symbols,
                                        node_text(&name_node, source).to_string(),
                                        type_name.to_string(),
                                    );
                                }
                            }
                        }
                    }
                }
            }
        }
        "formal_parameter" => {
            if let Some(type_node) = node.child_by_field_name("type") {
                if let Some(type_name) = extract_java_type_name(&type_node, source) {
                    if let Some(name_node) = node.child_by_field_name("name") {
                        push_type_map_entry(
                            symbols,
                            node_text(&name_node, source).to_string(),
                            type_name.to_string(),
                        );
                    }
                }
            }
        }
        _ => {}
    }
}

const JAVA_CLASS_KINDS: &[&str] = &["class_declaration", "enum_declaration", "interface_declaration"];

fn find_java_parent_class(node: &Node, source: &[u8]) -> Option<String> {
    find_enclosing_type_name(node, JAVA_CLASS_KINDS, source)
}

fn match_java_node(node: &Node, source: &[u8], symbols: &mut FileSymbols, _depth: usize) {
    match node.kind() {
        "class_declaration" => handle_class_decl(node, source, symbols),
        "interface_declaration" => handle_interface_decl(node, source, symbols),
        "enum_declaration" => handle_enum_decl(node, source, symbols),
        "method_declaration" => handle_method_decl(node, source, symbols),
        "constructor_declaration" => handle_constructor_decl(node, source, symbols),
        "import_declaration" => handle_import_decl(node, source, symbols),
        "method_invocation" => handle_method_invocation(node, source, symbols),
        "object_creation_expression" => handle_object_creation(node, source, symbols),
        _ => {}
    }
}

// ── Per-node-kind handlers for walk_node_depth ───────────────────────────────

fn handle_class_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let Some(name_node) = node.child_by_field_name("name") else { return };
    let class_name = node_text(&name_node, source).to_string();
    let children = extract_java_class_fields(node, source);
    symbols.definitions.push(Definition {
        name: class_name.clone(),
        kind: "class".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: opt_children(children),
    });

    // Superclass
    if let Some(superclass) = node.child_by_field_name("superclass") {
        extract_java_superclass(&superclass, &class_name, node, source, symbols);
    }

    // Interfaces
    if let Some(interfaces) = node.child_by_field_name("interfaces") {
        extract_java_interfaces(&interfaces, &class_name, source, symbols);
    }
}

fn extract_java_superclass(superclass: &Node, class_name: &str, node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    for i in 0..superclass.child_count() {
        let Some(child) = superclass.child(i) else { continue };
        match child.kind() {
            "type_identifier" | "identifier" => {
                symbols.classes.push(ClassRelation {
                    name: class_name.to_string(),
                    extends: Some(node_text(&child, source).to_string()),
                    implements: None,
                    line: start_line(node),
                });
                break;
            }
            "generic_type" => {
                if let Some(first) = child.child(0) {
                    symbols.classes.push(ClassRelation {
                        name: class_name.to_string(),
                        extends: Some(node_text(&first, source).to_string()),
                        implements: None,
                        line: start_line(node),
                    });
                }
                break;
            }
            _ => {}
        }
    }
}

fn handle_interface_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let Some(name_node) = node.child_by_field_name("name") else { return };
    let iface_name = node_text(&name_node, source).to_string();
    symbols.definitions.push(Definition {
        name: iface_name.clone(),
        kind: "interface".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: None,
    });
    if let Some(body) = node.child_by_field_name("body") {
        for i in 0..body.child_count() {
            let Some(child) = body.child(i) else { continue };
            if child.kind() != "method_declaration" { continue; }
            if let Some(meth_name) = child.child_by_field_name("name") {
                symbols.definitions.push(Definition {
                    name: format!("{}.{}", iface_name, node_text(&meth_name, source)),
                    kind: "method".to_string(),
                    line: start_line(&child),
                    end_line: Some(end_line(&child)),
                    decorators: None,
                    complexity: compute_all_metrics(&child, source, "java"),
                    cfg: build_function_cfg(&child, "java", source),
                    children: None,
                });
            }
        }
    }
}

fn handle_enum_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    if let Some(name_node) = node.child_by_field_name("name") {
        let enum_name = node_text(&name_node, source).to_string();
        let children = extract_java_enum_constants(node, source);
        symbols.definitions.push(Definition {
            name: enum_name,
            kind: "enum".to_string(),
            line: start_line(node),
            end_line: Some(end_line(node)),
            decorators: None,
            complexity: None,
            cfg: None,
            children: opt_children(children),
        });
    }
}

fn handle_method_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    // Skip interface methods — already emitted by handle_interface_decl without
    // parameters. Re-emitting them here would duplicate the definition and add
    // spurious `contains` edges to parameters of body-less declarations.
    // Mirrors C# handle_method_decl (csharp.rs) and the WASM Java extractor's
    // `node.parent?.parent?.type === 'interface_declaration'` guard.
    if let Some(parent) = node.parent() {
        if let Some(grand) = parent.parent() {
            if grand.kind() == "interface_declaration" { return; }
        }
    }
    if let Some(name_node) = node.child_by_field_name("name") {
        let parent_class = find_java_parent_class(node, source);
        let name = node_text(&name_node, source);
        let full_name = match &parent_class {
            Some(cls) => format!("{}.{}", cls, name),
            None => name.to_string(),
        };
        let children = extract_java_parameters(node, source);
        symbols.definitions.push(Definition {
            name: full_name,
            kind: "method".to_string(),
            line: start_line(node),
            end_line: Some(end_line(node)),
            decorators: None,
            complexity: compute_all_metrics(node, source, "java"),
            cfg: build_function_cfg(node, "java", source),
            children: opt_children(children),
        });
    }
}

fn handle_constructor_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    if let Some(name_node) = node.child_by_field_name("name") {
        let parent_class = find_java_parent_class(node, source);
        let name = node_text(&name_node, source);
        let full_name = match &parent_class {
            Some(cls) => format!("{}.{}", cls, name),
            None => name.to_string(),
        };
        let children = extract_java_parameters(node, source);
        symbols.definitions.push(Definition {
            name: full_name,
            kind: "method".to_string(),
            line: start_line(node),
            end_line: Some(end_line(node)),
            decorators: None,
            complexity: compute_all_metrics(node, source, "java"),
            cfg: build_function_cfg(node, "java", source),
            children: opt_children(children),
        });
    }
}

fn handle_import_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let mut import_path = String::new();
    let mut has_asterisk = false;
    for i in 0..node.child_count() {
        if let Some(child) = node.child(i) {
            if child.kind() == "scoped_identifier" || child.kind() == "identifier" {
                import_path = node_text(&child, source).to_string();
            }
            if child.kind() == "asterisk" {
                has_asterisk = true;
            }
        }
    }
    if !import_path.is_empty() {
        let names = if has_asterisk {
            vec!["*".to_string()]
        } else {
            let last = import_path.split('.').last().unwrap_or("").to_string();
            vec![last]
        };
        push_import(symbols, node, import_path, names, |imp| {
            imp.java_import = Some(true);
        });
    }
}

fn handle_method_invocation(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    if let Some(name_node) = node.child_by_field_name("name") {
        let receiver = named_child_text(node, "object", source)
            .map(|s| s.to_string());
        push_call(
            symbols,
            node,
            node_text(&name_node, source).to_string(),
            receiver,
            None,
        );
    }
}

fn handle_object_creation(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let Some(type_node) = node.child_by_field_name("type") else { return };
    let type_name = if type_node.kind() == "generic_type" {
        type_node.child(0).map(|n| node_text(&n, source).to_string())
    } else {
        Some(node_text(&type_node, source).to_string())
    };
    if let Some(name) = type_name {
        push_simple_call(symbols, node, name);
    }
}

// ── Extended kinds helpers ──────────────────────────────────────────────────

fn extract_java_parameters(node: &Node, source: &[u8]) -> Vec<Definition> {
    let params_node = node
        .child_by_field_name("parameters")
        .or_else(|| find_child(node, "formal_parameters"));
    extract_simple_parameters(
        params_node,
        source,
        &ExtractParametersOptions {
            param_kinds: &["formal_parameter", "spread_parameter"],
            name_field: Some("name"),
            fallback_to_identifier: false,
        },
    )
}

fn extract_java_class_fields(node: &Node, source: &[u8]) -> Vec<Definition> {
    let mut fields = Vec::new();
    let body = node.child_by_field_name("body");
    if let Some(body) = body {
        for i in 0..body.child_count() {
            if let Some(child) = body.child(i) {
                if child.kind() == "field_declaration" {
                    // Field declarators contain the names
                    for j in 0..child.child_count() {
                        if let Some(decl) = child.child(j) {
                            if decl.kind() == "variable_declarator" {
                                if let Some(name_node) = decl.child_by_field_name("name") {
                                    fields.push(child_def(
                                        node_text(&name_node, source).to_string(),
                                        "property",
                                        start_line(&child),
                                    ));
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    fields
}

fn extract_java_enum_constants(node: &Node, source: &[u8]) -> Vec<Definition> {
    let mut constants = Vec::new();
    let body = node.child_by_field_name("body");
    if let Some(body) = body {
        for i in 0..body.child_count() {
            if let Some(child) = body.child(i) {
                if child.kind() == "enum_constant" {
                    if let Some(name_node) = child.child_by_field_name("name") {
                        constants.push(child_def(
                            node_text(&name_node, source).to_string(),
                            "constant",
                            start_line(&child),
                        ));
                    }
                }
            }
        }
    }
    constants
}

// ── Existing helpers ────────────────────────────────────────────────────────

fn extract_java_interfaces(
    interfaces: &Node,
    class_name: &str,
    source: &[u8],
    symbols: &mut FileSymbols,
) {
    for i in 0..interfaces.child_count() {
        let Some(child) = interfaces.child(i) else { continue };
        match child.kind() {
            "type_identifier" | "identifier" => {
                push_implements(symbols, class_name, node_text(&child, source), interfaces);
            }
            "type_list" => {
                collect_type_list_implements(&child, class_name, source, symbols, interfaces);
            }
            "generic_type" => {
                if let Some(first) = child.child(0) {
                    push_implements(symbols, class_name, node_text(&first, source), interfaces);
                }
            }
            _ => {}
        }
    }
}

fn push_implements(symbols: &mut FileSymbols, class_name: &str, iface_name: &str, line_node: &Node) {
    symbols.classes.push(ClassRelation {
        name: class_name.to_string(),
        extends: None,
        implements: Some(iface_name.to_string()),
        line: start_line(line_node),
    });
}

fn collect_type_list_implements(
    type_list: &Node,
    class_name: &str,
    source: &[u8],
    symbols: &mut FileSymbols,
    line_node: &Node,
) {
    for j in 0..type_list.child_count() {
        let Some(t) = type_list.child(j) else { continue };
        match t.kind() {
            "type_identifier" | "identifier" => {
                push_implements(symbols, class_name, node_text(&t, source), line_node);
            }
            "generic_type" => {
                if let Some(first) = t.child(0) {
                    push_implements(symbols, class_name, node_text(&first, source), line_node);
                }
            }
            _ => {}
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tree_sitter::Parser;

    fn parse_java(code: &str) -> FileSymbols {
        let mut parser = Parser::new();
        parser
            .set_language(&tree_sitter_java::LANGUAGE.into())
            .unwrap();
        let tree = parser.parse(code.as_bytes(), None).unwrap();
        JavaExtractor.extract(&tree, code.as_bytes(), "Test.java")
    }

    // ── Extended kinds tests ────────────────────────────────────────────────

    #[test]
    fn extracts_method_parameters() {
        let s = parse_java("class Foo { void bar(int x, String y) {} }");
        let bar = s.definitions.iter().find(|d| d.name == "Foo.bar").unwrap();
        let children = bar.children.as_ref().unwrap();
        assert_eq!(children.len(), 2);
        assert_eq!(children[0].name, "x");
        assert_eq!(children[0].kind, "parameter");
        assert_eq!(children[1].name, "y");
    }

    #[test]
    fn extracts_class_fields() {
        let s = parse_java("class User { String name; int age; }");
        let user = s.definitions.iter().find(|d| d.name == "User").unwrap();
        let children = user.children.as_ref().unwrap();
        let names: Vec<&str> = children.iter().map(|c| c.name.as_str()).collect();
        assert!(names.contains(&"name"));
        assert!(names.contains(&"age"));
        assert!(children.iter().all(|c| c.kind == "property"));
    }

    #[test]
    fn extracts_enum_constants() {
        let s = parse_java("enum Status { ACTIVE, INACTIVE }");
        let status = s.definitions.iter().find(|d| d.name == "Status").unwrap();
        let children = status.children.as_ref().unwrap();
        assert_eq!(children.len(), 2);
        assert_eq!(children[0].name, "ACTIVE");
        assert_eq!(children[0].kind, "constant");
        assert_eq!(children[1].name, "INACTIVE");
    }
}

File: src/extractors/java.ts
import type {
  ExtractorOutput,
  SubDeclaration,
  TreeSitterNode,
  TreeSitterTree,
  TypeMapEntry,
} from '../types.js';
import {
  extractBodyMembers,
  extractModifierVisibility,
  extractSimpleParameters,
  findChild,
  findParentNode,
  lastPathSegment,
  nodeEndLine,
  nodeStartLine,
  pushCall,
  pushImport,
  setTypeMapEntry,
} from './helpers.js';

/**
 * Extract symbols from Java files.
 */
export function extractJavaSymbols(tree: TreeSitterTree, _filePath: string): ExtractorOutput {
  const ctx: ExtractorOutput = {
    definitions: [],
    calls: [],
    imports: [],
    classes: [],
    exports: [],
    typeMap: new Map(),
  };

  walkJavaNode(tree.rootNode, ctx);
  return ctx;
}

function walkJavaNode(node: TreeSitterNode, ctx: ExtractorOutput): void {
  switch (node.type) {
    case 'class_declaration':
      handleJavaClassDecl(node, ctx);
      break;
    case 'interface_declaration':
      handleJavaInterfaceDecl(node, ctx);
      break;
    case 'enum_declaration':
      handleJavaEnumDecl(node, ctx);
      break;
    case 'method_declaration':
      handleJavaMethodDecl(node, ctx);
      break;
    case 'constructor_declaration':
      handleJavaConstructorDecl(node, ctx);
      break;
    case 'import_declaration':
      handleJavaImportDecl(node, ctx);
      break;
    case 'method_invocation':
      handleJavaMethodInvocation(node, ctx);
      break;
    case 'object_creation_expression':
      handleJavaObjectCreation(node, ctx);
      break;
    case 'local_variable_declaration':
      handleJavaLocalVarDecl(node, ctx);
      break;
  }

  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (child) walkJavaNode(child, ctx);
  }
}

// ── Walk-path per-node-type handlers ────────────────────────────────────────

function handleJavaClassDecl(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  const classChildren = extractClassFields(node);
  ctx.definitions.push({
    name: nameNode.text,
    kind: 'class',
    line: nodeStartLine(node),
    endLine: nodeEndLine(node),
    children: classChildren.length > 0 ? classChildren : undefined,
  });

  extractJavaSuperclass(node, nameNode.text, ctx);

  const interfaces = node.childForFieldName('interfaces');
  if (interfaces) {
    extractJavaInterfaces(interfaces, nameNode.text, nodeStartLine(node), ctx);
  }
}

/** Extract the superclass (extends) relationship from a Java class declaration. */
function extractJavaSuperclass(
  node: TreeSitterNode,
  className: string,
  ctx: ExtractorOutput,
): void {
  const superclass = node.childForFieldName('superclass');
  if (!superclass) return;
  const superName = findJavaSuperTypeName(superclass);
  if (superName) {
    ctx.classes.push({ name: className, extends: superName, line: nodeStartLine(node) });
  }
}

/** Find the type name from a superclass node (handles generic_type unwrapping). */
function findJavaSuperTypeName(superclass: TreeSitterNode): string | undefined {
  for (let i = 0; i < superclass.childCount; i++) {
    const child = superclass.child(i);
    if (!child) continue;
    if (JAVA_TYPE_NODE_TYPES.has(child.type)) {
      return resolveJavaIfaceName(child);
    }
  }
  return undefined;
}

const JAVA_TYPE_NODE_TYPES = new Set(['type_identifier', 'identifier', 'generic_type']);

/** Resolve interface name from a type node (handles generic_type unwrapping). */
function resolveJavaIfaceName(node: TreeSitterNode): string | undefined {
  return node.type === 'generic_type' ? node.child(0)?.text : node.text;
}

/** Push a single interface type node as an implements entry. */
function pushJavaIface(
  node: TreeSitterNode,
  className: string,
  line: number,
  ctx: ExtractorOutput,
): void {
  if (!JAVA_TYPE_NODE_TYPES.has(node.type)) return;
  const ifaceName = resolveJavaIfaceName(node);
  if (ifaceName) ctx.classes.push({ name: className, implements: ifaceName, line });
}

function extractJavaInterfaces(
  interfaces: TreeSitterNode,
  className: string,
  line: number,
  ctx: ExtractorOutput,
): void {
  for (let i = 0; i < interfaces.childCount; i++) {
    const child = interfaces.child(i);
    if (!child) continue;

    if (child.type === 'type_list') {
      for (let j = 0; j < child.childCount; j++) {
        const t = child.child(j);
        if (t) pushJavaIface(t, className, line, ctx);
      }
    } else {
      pushJavaIface(child, className, line, ctx);
    }
  }
}

function handleJavaInterfaceDecl(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  ctx.definitions.push({
    name: nameNode.text,
    kind: 'interface',
    line: nodeStartLine(node),
    endLine: nodeEndLine(node),
  });
  const body = node.childForFieldName('body');
  if (body) extractJavaInterfaceMethods(body, nameNode.text, ctx);
}

/** Extract method declarations from a Java interface body. */
function extractJavaInterfaceMethods(
  body: TreeSitterNode,
  ifaceName: string,
  ctx: ExtractorOutput,
): void {
  for (let i = 0; i < body.childCount; i++) {
    const child = body.child(i);
    if (child && child.type === 'method_declaration') {
      const methName = child.childForFieldName('name');
      if (methName) {
        ctx.definitions.push({
          name: `${ifaceName}.${methName.text}`,
          kind: 'method',
          line: nodeStartLine(child),
          endLine: nodeEndLine(child),
        });
      }
    }
  }
}

function handleJavaEnumDecl(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  const enumChildren = extractEnumConstants(node);
  ctx.definitions.push({
    name: nameNode.text,
    kind: 'enum',
    line: nodeStartLine(node),
    endLine: nodeEndLine(node),
    children: enumChildren.length > 0 ? enumChildren : undefined,
  });
}

function handleJavaMethodDecl(node: TreeSitterNode, ctx: ExtractorOutput): void {
  // Skip interface methods already emitted by handleJavaInterfaceDecl
  if (node.parent?.parent?.type === 'interface_declaration') return;
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  const parentClass = findJavaParentClass(node);
  const fullName = parentClass ? `${parentClass}.${nameNode.text}` : nameNode.text;
  const params = extractJavaParameters(node.childForFieldName('parameters'), ctx.typeMap);
  ctx.definitions.push({
    name: fullName,
    kind: 'method',
    line: nodeStartLine(node),
    endLine: nodeEndLine(node),
    children: params.length > 0 ? params : undefined,
    visibility: extractModifierVisibility(node),
  });
}

function handleJavaConstructorDecl(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  const parentClass = findJavaParentClass(node);
  const fullName = parentClass ? `${parentClass}.${nameNode.text}` : nameNode.text;
  const params = extractJavaParameters(node.childForFieldName('parameters'), ctx.typeMap);
  ctx.definitions.push({
    name: fullName,
    kind: 'method',
    line: nodeStartLine(node),
    endLine: nodeEndLine(node),
    children: params.length > 0 ? params : undefined,
    visibility: extractModifierVisibility(node),
  });
}

function handleJavaImportDecl(node: TreeSitterNode, ctx: ExtractorOutput): void {
  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (child && (child.type === 'scoped_identifier' || child.type === 'identifier')) {
      const fullPath = child.text;
      const lastName = lastPathSegment(fullPath, '.');
      pushImport(ctx, node, fullPath, [lastName], { javaImport: true });
    }
    if (child && child.type === 'asterisk') {
      const lastImport = ctx.imports[ctx.imports.length - 1];
      if (lastImport) lastImport.names = ['*'];
    }
  }
}

function handleJavaMethodInvocation(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  const obj = node.childForFieldName('object');
  pushCall(ctx, node, nameNode.text, obj ? { receiver: obj.text } : {});
}

function handleJavaLocalVarDecl(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const typeNode = node.childForFieldName('type');
  if (!typeNode) return;
  const typeName = resolveJavaTypeText(typeNode);
  if (!typeName) return;
  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (child?.type === 'variable_declarator') {
      const nameNode = child.childForFieldName('name');
      // Use setTypeMapEntry (first-wins on tie) to match Rust extractor semantics.
      // The typeMap is flat per-file without method scoping, so a local variable
      // in one method (e.g. `InMemoryUserRepository repo` in `createDefault()`) must
      // not override a parameter binding set by an earlier method
      // (e.g. `UserRepository repo` constructor param). First-wins preserves the
      // interface/abstract type annotation that drives correct CHA dispatch.
      if (nameNode && ctx.typeMap) setTypeMapEntry(ctx.typeMap, nameNode.text, typeName, 0.9);
    }
  }
}

function handleJavaObjectCreation(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const typeNode = node.childForFieldName('type');
  if (!typeNode) return;
  const typeName = resolveJavaTypeText(typeNode);
  if (typeName) pushCall(ctx, node, typeName);
}

/**
 * Resolve a Java type node's text, unwrapping `generic_type` to its base name.
 * Used wherever we need the bare type identifier (local var decls, object
 * creation, parameter types).
 */
function resolveJavaTypeText(typeNode: TreeSitterNode): string | undefined {
  return typeNode.type === 'generic_type' ? typeNode.child(0)?.text : typeNode.text;
}

const JAVA_PARENT_TYPES = [
  'class_declaration',
  'enum_declaration',
  'interface_declaration',
] as const;
function findJavaParentClass(node: TreeSitterNode): string | null {
  return findParentNode(node, JAVA_PARENT_TYPES);
}

// ── Child extraction helpers ────────────────────────────────────────────────

const JAVA_PARAM_TYPES = ['formal_parameter', 'spread_parameter'] as const;

function extractJavaParameters(
  paramListNode: TreeSitterNode | null,
  typeMap?: Map<string, TypeMapEntry>,
): SubDeclaration[] {
  return extractSimpleParameters(paramListNode, {
    paramTypes: JAVA_PARAM_TYPES,
    typeMap,
    resolveType: resolveJavaTypeText,
  });
}

function extractClassFields(classNode: TreeSitterNode): SubDeclaration[] {
  const fields: SubDeclaration[] = [];
  const body = classNode.childForFieldName('body') || findChild(classNode, 'class_body');
  if (!body) return fields;
  for (let i = 0; i < body.childCount; i++) {
    const member = body.child(i);
    if (member?.type !== 'field_declaration') continue;
    extractFieldDeclarators(member, fields);
  }
  return fields;
}

/** Extract variable_declarator names from a field_declaration node. */
function extractFieldDeclarators(member: TreeSitterNode, fields: SubDeclaration[]): void {
  const vis = extractModifierVisibility(member);
  for (let j = 0; j < member.childCount; j++) {
    const child = member.child(j);
    if (child?.type !== 'variable_declarator') continue;
    const nameNode = child.childForFieldName('name');
    if (nameNode) {
      fields.push({
        name: nameNode.text,
        kind: 'property',
        line: nodeStartLine(member),
        visibility: vis,
      });
    }
  }
}

function extractEnumConstants(enumNode: TreeSitterNode): SubDeclaration[] {
  return extractBodyMembers(enumNode, ['body', 'enum_body'], 'enum_constant', 'constant');
}
